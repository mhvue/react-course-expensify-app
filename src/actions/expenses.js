import database from "../firebase/firebase";
//actions:
//ADD EXPENSE
export const addExpense = (expense) => ({
    type: "ADD_EXPENSE",
    expense 
});
//REMOVE EXPENSE
export const removeExpense = ({id} = {}) => ({
    type: "REMOVE_EXPENSE",
    id
});
//EDIT EXPENSE
export const editExpense = (id, updates) => ({
    type: "EDIT_EXPENSE",
    id,
    updates
});

export const startAddExpense = (expenseData = {}) => {
    return (dispatch) => {
        const {
            description = "", 
            note = "", 
            amount = 0, 
            createdAt=0
        } = expenseData;
        //access firebase and save data 

        const expense = {
            description, 
            note, 
            amount, 
            createdAt
        }
        return database.ref("expenses").push(expense)
            .then((ref) => {
            dispatch(addExpense({
                id: ref.key,
                ...expense
            }));
        });
    }
}