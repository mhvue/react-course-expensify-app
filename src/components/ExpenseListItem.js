import React from "react";
import { Link } from "react-router-dom";
import moment from "moment";
import numeral from "numeral";

const ExpenseItemList = ({id, description, createdAt, amount}) =>  (
    <div>
    <Link to={`/edit/${id}`}>   
      <p>{description}</p>
    </Link>
      <p>Created at: {moment(createdAt).format("MMMM Do, YYYY")} </p>
       <p>Amount: {numeral(amount / 100).format("$0,0.00")}</p>     
    </div>
);


export default ExpenseItemList;