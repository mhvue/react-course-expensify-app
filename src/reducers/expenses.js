
//Expenses reducer 
const expensesReducerDefaultState = [];

export default (state = expensesReducerDefaultState, action) => {
    switch(action.type){
        case "ADD_EXPENSE" : 
            return [
                ...state, //this is all the our current items in array 
                action.expense
            ];
        case "REMOVE_EXPENSE": 
            return state.filter(({id}) => {
                return id !== action.id
            });
        case "EDIT_EXPENSE" : 
            return state.map((expense) => {
                if(expense.id === action.id){
                    return {
                        ...expense, //grab all expenses
                        ...action.updates
                    }
                }else{
                    return expense;
                }
            })
        case "SET_EXPENSES": 
            return action.expense;
        default: 
            return state;
    }
};

