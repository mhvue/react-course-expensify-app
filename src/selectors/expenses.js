import moment from "moment";
//get visibile expenses
export default (expenses, { text, sortBy, startDate, endDate }) => {
    return expenses.filter((expense) => {
      const createdAtMoment = moment(expense.createdAt);

      const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, "day") : true;
      const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, "day"): true;
     const textMatch = expense.description.toLowerCase().includes(text.toLowerCase())
     console.log(expense)

      return startDateMatch && endDateMatch && textMatch;

   }).sort((a,b) => {
    if(sortBy === "date") {
        //return -1 if "a" should come first and return 1 if "b" should come first 
        return a.createdAt < b.createdAt ? 1  : -1;
    }
    else if(sortBy === "amount") {
        return a.amount  < b.amount ? 1 : -1;
    }
   });
  };