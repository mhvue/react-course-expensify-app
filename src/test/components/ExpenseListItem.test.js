import React from "react";
import { shallow } from "enzyme";
import ExpenseListItem  from "../../components/ExpenseListItem";
import expenses from "../fixtures/expenses";

test("should render ExpenseListItem with each expense", () => {
    const wrapper = shallow(<ExpenseListItem 
                                id={expenses[1].id}
                                description={expenses[1].description}
                                createdAt={expenses[1].createdAt}
                                amount={expenses[1].amount}
                            />)
    expect(wrapper).toMatchSnapshot();
})