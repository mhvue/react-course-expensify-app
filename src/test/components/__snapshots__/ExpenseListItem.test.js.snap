// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseListItem with each expense 1`] = `
<div>
  <Link
    replace={false}
    to="/edit/2"
  >
    <p>
      Rent
    </p>
  </Link>
  <p>
    Created at: 
    December 27th, 1969
     
  </p>
  <p>
    Amount: 
    $1,095.00
  </p>
</div>
`;
